src/app/modules/form/admin-form/admin-form.service.ts

```diff
@@ -1148,3 +1164,172 @@ const isMobileFieldUpdateAllowed = (
     },
   )
 }
+
+export const createTwilioCredentials = (
+  twilioCredentials: TwilioCredentials,
+  form: IPopulatedForm,
+): ResultAsync<unknown, ApplicationError> => {
+  if (!isCredentialsValid(twilioCredentials))
+    return errAsync(new MalformedParametersError('Credentials are invalid!'))
+
+  const twilioCredentialsData: TwilioCredentialsData =
+    new TwilioCredentialsData(twilioCredentials)
+
+  const msgSrvcName = `formsg/${process.env.SSM_PREFIX}/form/${form._id}/twilio`
+  const body: SecretsManager.Types.CreateSecretRequest = {
+    Name: msgSrvcName,
+    SecretString: twilioCredentialsData.toString(),
+  }
+  logger.info({
+    message: `no msgSrvcName, creating Twilio credentials for form ${form._id}`,
+    meta: {
+      action: 'createTwilioCredentials',
+      formId: form._id.toHexString(),
+      msgSrvcName,
+    },
+  })
+
+  return ResultAsync.fromPromise(
+    mongoose.connection.transaction(async (session: ClientSession) => {
+      await FormModel.updateMsgSrvcName(form._id, msgSrvcName, session)
+      await secretsManager.createSecret(body).promise()
+    }),
+    (error) => {
+      logger.error({
+        message: 'Error encountered when creating Twilio Secret',
+        meta: {
+          action: 'createTwilio',
+          body,
+        },
+        error,
+      })
+
+      return new ApplicationError('Error occurred when creating Twilio!')
+    },
+  )
+}
+
+export const updateTwilioCredentials = (
+  msgSrvcName: string,
+  twilioCredentials: TwilioCredentials,
+): ResultAsync<PutSecretValueResponse, DatabaseError> => {
+  if (!isCredentialsValid(twilioCredentials))
+    return errAsync(new MalformedParametersError('Credentials are invalid!'))
+
+  const twilioCredentialsData: TwilioCredentialsData =
+    new TwilioCredentialsData(twilioCredentials)
+
+  const body: SecretsManager.Types.PutSecretValueRequest = {
+    SecretId: msgSrvcName,
+    SecretString: twilioCredentialsData.toString(),
+  }
+
+  return ResultAsync.fromPromise(
+    secretsManager.getSecretValue({ SecretId: msgSrvcName }).promise(),
+    (error) => {
+      logger.error({
+        message: 'Twilio Credentials do not exist in Secrets Manager',
+        meta: {
+          action: 'updateTwilioCredentials',
+          msgSrvcName,
+        },
+        error,
+      })
+
+      return new DatabaseError(
+        'Twilio Credentials do not exist in Secrets Manager',
+      )
+    },
+  ).andThen(() => {
+    logger.info({
+      message: 'Twilio Credentials has been found in Secrets Manager',
+      meta: {
+        action: 'updateTwilioCredentials',
+        msgSrvcName,
+      },
+    })
+
+    twilioClientCache.del(msgSrvcName)
```

Comment:
we may want to do this AFTER we update the value, or we may get another cache hit before we update.

alternatively, we may not even need to bust the cache, you may want to document why we do so here

Reply:
Updated and documented the required change, currently, if a twilio credentials retrieve call happens within 10s of updating secrets manager, it will be a cache hit and return the wrong twilio credential